Kdump support

arnaldo2792 opened the pull request:
**Issue number:**
#1413 

**Description of changes:**

```
3883f7f7 os: Add kdump support
842604d7 kernel-5.4: add patches required for kdump support
c78ec8a2 systemd: move systemd mounts to preconfigured.target
1e0f6859 packages: add makedumpfile
1466a49f packages: add libelf
bcc5c903 packages: add kexec-tools
```
Kdump is a Linux feature that allows to boot to a kernel whenever the system panics. The crash kernel is loaded into a reserved space in memory determined by the `crashkernel` kernel parameter. In Bottlerocket, this parameter is set such that no memory will be reserved if the host has less than 2GB of memory.

For Bottlerocket, the crash kernel is loaded from the current active boot partition. The `configure-boot-mount.service` systemd unit determines which is the current active boot partition and mounts it at `/boot`. This mount is set as read-only and with private propagations, so new mount namespaces won't have access to it. As part of this change, SELinux labels are added to the boot partition when it is created by the `rpm2img` tool.

The `load-crash-kernel.service` systemd unit loads the crash kernel, only if memory was reserved for it, and the `kexec.kexec_load_disable` setting is `0`. The unit will exit gracefully if no memory was reserved for the crash kernel. For the moment only the aws-dev and vmware variants use that kernel parameter.

The `kexec.kexec_load_disable` setting used to be set in the `sysctl.conf` configuration file. With this change, the setting is set using the `disable-kexec-load.service` systemd unit. This unit runs after `load-crash-kernel.service`, even if the latter wasn't executed or excited with a non-zero code.

The `capture-kernel-dump.service` systemd unit is set as the target when the crash kernel is executed. It captures both the dmesg logs and the kdump-compressed dump excluding:
      * Pages filled with zero
      * Non-private cache pages
      * All cache pages
      * User process data pages
      * Free pages

All the files generated by the `capture-kernel-dump.service` unit are stored at `/var/log/kdump`, therefore the unit has a strong dependency on the following services to setup the persistent partition:
  * local-fs.target
  * systemd-sysusers.service
  * systemd-udevd.service
  * systemd-udev-trigger.service
  * systemd-tmpfiles-setup.service
  * systemd-tmpfiles-setup-dev.service

Since `local-fs.target` is a dependency of `capture-kernel-dump.service`, systemd will attempt to load all the mount units. To prevent this, the mount units will only be loaded during the execution of the `preconfigured` target.

No API is provided to enable/disable the dump collection, since the memory space is reserved and it will be a waste if nothing uses that space. Dynamically changing the `crashkernel` cmd line parameter isn't an option since we will provide support for secure boot in the future.

**Testing done:**
aws-dev x86_64/aarch64, vmware-dev/vmware-k8s-1.20 x86_64:

- `systemctl status` didn't show failed units
- Crashed the kernel with `echo c > /proc/sysrq-trigger`, and verified that the dumps/logs were generated
- Verified that existing dump files were deleted
- Verified that the correct active boot partition was mounted

k8s variant x86_64:

- I did a stress test to verify that `256MB` of memory are enough to collect the kernel dumps, regardless of the size of the host and the operation workload it has. I used a custom build for the aws-k8s-1.20 variant for this test and a `m5.8xlarge` EC2 instance with 110 pods running that only load random data and keep it there forever. 
- During the same test, max out the number of volumes that can be attached to an instance, since udev was causing OOM errors before I limited the number of child processes that it can have. I verified that the dumps were generated properly during these tests:

```
[ec2-user@ip-192-168-0-174 log]$ free -h
              total        used        free      shared  buff/cache   available
Mem:           122G        112G        7.2G         18M        2.8G        8.9G
Swap:            0B          0B          0B
```

```
-rw-------. 1 root root  59K May 21 00:14 dmesg.dump
-rw-------. 1 root root 790M May 21 00:14 kdump.dump
-rw-r--r--. 1 root root   77 May 21 00:14 prairie-dog.log
```
aws-ecs-1, aws-k8s-1.19 x86_64:
- `systemctl status` didn't show failed units
- Run nginx task/pod
- Validated that `kexec.service` wasn't executed:

```
● kexec.service - Load crash kernel
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/kexec.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
  Condition: start condition failed at Tue 2021-06-01 21:38:07 UTC; 17min ago
             └─ ConditionKernelCommandLine=crashkernel was not met

Jun 01 21:38:04 localhost systemd[1]: Condition check resulted in Load crash kernel being skipped.
Jun 01 21:38:07 ip-192-168-72-37.us-west-2.compute.internal systemd[1]: Condition check resulted in Load crash kernel being skipped.
Jun 01 21:38:07 ip-192-168-72-37.us-west-2.compute.internal systemd[1]: Condition check resulted in Load crash kernel being skipped.
```

Custom aws-k8s-1.19 x86_64 build with `crashkernel`, to validate the 5.4 kernel behavior:
- `systemctl status` didn't show failed units
- Run nginx pod
- Crashed the kernel with `echo c > /proc/sysrq-trigger`, and verified that the dumps/logs were generated

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
packages: add kexec-tools

This commit adds kexec-tools to all variants

arnaldo2792 committed:
packages: add libelf

This commit adds libelf to all variants

arnaldo2792 committed:
packages: add makedumpfile

This commit adds makedumpfile to all variants

arnaldo2792 committed:
os: Add kdump support

Kdump is a Linux feature that allows to boot to a kernel whenever the
system panics. The crash kernel is loaded into a reserved space in
memory determined by the `crashkernel` kernel parameter. In
Bottlerocket, this parameter is set such that no memory will be reserved
if the host has less than 2GB of memory.

For Bottlerocket, the crash kernel is loaded from the current active
boot partition. The `configure-boot-mount.service` systemd unit
determines which is the current active boot partition and mounts it at
`/boot`. This mount is set as read-only and with private propagations,
so new mount namespaces won't have access to it. As part of this change,
SELinux labels are added to the boot partition when it is created by the
`rpm2img` tool.

The `load-crash-kernel.service` systemd unit loads the crash kernel,
only if memory was reserved for it, and the `kexec.kexec_load_disable`
setting is `0`. The unit will exit gracefully if no memory was reserved
for the crash kernel. For the moment only the aws-dev and vmware variants
use that kernel parameter.

The `kexec.kexec_load_disable` setting used to be set in the
`sysctl.conf` configuration file. With this change, the setting is set
using the `disable-kexec-load.service` systemd unit. This unit runs after
`load-crash-kernel.service`, even if the latter wasn't executed or
excited with a non-zero code.

The `capture-kernel-dump.service` systemd unit is set as the target when
the crash kernel is executed. It captures both the dmesg logs and the
kdump-compressed dump excluding:
  * Pages filled with zero
  * Non-private cache pages
  * All cache pages
  * User process data pages
  * Free pages

All the files generated by the `capture-kernel-dump.service` unit are
stored at `/var/log/kdump`, therefore the unit has a strong dependency
on the following services to set up the persistent partition:
  * local-fs.target
  * systemd-sysusers.service
  * systemd-udevd.service
  * systemd-udev-trigger.service
  * systemd-tmpfiles-setup.service
  * systemd-tmpfiles-setup-dev.service

Since `local-fs.target` is a dependency of `capture-kernel-dump.service`,
systemd will attempt to load all the mount units. To prevent this, the
mount units will only be loaded during the execution of the `preconfigured`
target.

No API is provided to enable/disable the dump collection, since the
memory space is reserved and it will be a waste if nothing uses that
space. Dynamically changing the `crashkernel` cmd line parameter isn't
an option since we will provide support for secure boot in the future.

arnaldo2792 commented:
- Update the URL for libelf's spec file

arnaldo2792 commented:
- Remove leftover systemd service

bcressey commented on packages/kexec-tools/Cargo.lock:
```diff
@@ -0,0 +1,5 @@
+# It is not intended for manual editing.
+[[package]]
+name = "kexec-tools"
+version = "0.1.0"
```
We shouldn't need this file.

bcressey commented on packages/kexec-tools/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+
+[build-dependencies]
+glibc = { path = "../glibc" }
+ncurses = { path = "../ncurses" }
```
Why do we need ncurses?

bcressey commented on packages/kexec-tools/kexec-tools.spec:
```diff
@@ -0,0 +1,32 @@
+Version: 2.0.21
+Release: 1%{?dist}
+Summary: Linux tool to load kernels from the running system
+License: GPL-2.0
```
This is a deprecated SPDX identifier, and there's a mix of both licenses.

```suggestion
License: GPL-2.0-or-later AND GPL-2.0-only
```

bcressey commented on packages/kexec-tools/kexec-tools.spec:
```diff
@@ -0,0 +1,32 @@
+
+%build
+%cross_configure
+rm -f kexec-tools.spec.in
```
Do we need this? Generally if we are making changes to unpacked sources, that would happen in `%prep`.

bcressey commented on packages/kexec-tools/kexec-tools.spec:
```diff
@@ -0,0 +1,32 @@
+%build
+%cross_configure
+rm -f kexec-tools.spec.in
+make
```
Does this not build correctly with multiple jobs?

```suggestion
%make_build
```

bcressey commented on packages/libelf/libelf.spec:
```diff
@@ -0,0 +1,57 @@
+Name: %{_cross_os}libelf
+Version: 0.183
```
Why `0.183` instead of `0.184` or `0.185`?

bcressey commented on packages/libelf/libelf.spec:
```diff
@@ -0,0 +1,57 @@
+
+%package devel
+Summary: Files for development using the library for EFL files
+License: GPL-2.0-or-later OR LGPL-3.0-or-later
```
Don't need to repeat the license if it's the same.

```suggestion
```

bcressey commented on packages/libelf/libelf.spec:
```diff
@@ -0,0 +1,57 @@
+%{summary}.
+
+%package devel
+Summary: Files for development using the library for EFL files
```
```suggestion
Summary: Files for development using the library for ELF files
```

bcressey commented on packages/libelf/libelf.spec:
```diff
@@ -0,0 +1,57 @@
+Name: %{_cross_os}libelf
+Version: 0.183
+Release: 1%{?dist}
+Summary: Library to handle ELF files
```
```suggestion
Summary: Library for ELF files
```

bcressey commented on packages/makedumpfile/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+
+[dependencies]
+libelf = { path = "../libelf" }
+libz = { path = "../libz" }
```
nit: we haven't been including dependencies that overlap with build-dependencies.
```suggestion
```

bcressey commented on packages/makedumpfile/makedumpfile.spec:
```diff
@@ -0,0 +1,50 @@
+Patch9001: 9001-printk-add-support-for-lockless-ringbuffer.patch
+Patch9002: 9002-printk-use-committed-finalized-state-values.patch
+
+BuildRequires: %{_cross_os}libbzip2-devel
```
Do we need to enable bzip2 support? I'd prefer not to, but otherwise I'd expect it to show up as a runtime requires.

arnaldo2792 replied:
We need to add the dependency otherwise the build fails, I added runtime requires to the spec and the toml files

bcressey commented on packages/makedumpfile/makedumpfile.spec:
```diff
@@ -0,0 +1,50 @@
+Version: 1.6.8
+Release: 1%{?dist}
+Summary: Tool to create dumps from kernel memory images
+License: GPL-2.0-or-later
```
There are some files like `arch/ppc.c` that don't have the "any later version" text.

```suggestion
License: GPL-2.0-or-later AND GPL-2.0-only
```

bcressey commented on packages/os/Cargo.toml:
```diff
@@ -13,6 +13,7 @@ source-groups = [
     "metricdog",
     "parse-datetime",
     "ghostdog",
+    "prairie-dog",
```
`prairiedog` is more in line with our naming convention, such as it is 😄

bcressey commented on packages/os/os.spec:
```diff
@@ -514,6 +545,15 @@ install -p -m 0644 %{S:300} %{buildroot}%{_cross_udevrulesdir}/80-ephemeral-stor
 %{_cross_bindir}/static-pods
 %endif
 
+%if %{_is_vmware_variant} || %{_is_dev_variant}
+%files -n %{_cross_os}prairie-dog
+%{_cross_bindir}/prairie-dog
+%{_cross_unitdir}/kdump.service
+%{_cross_unitdir}/kexec.service
+%{_cross_unitdir}/configure-boot-mount.service
+%{_cross_tmpfilesdir}/kdump.conf
+%endif
```
I'd prefer to unconditionally build and include `prairiedog` and decide at runtime whether to do anything based on the presence of `crashkernel` in the command line.

Possibly this could use the `ConditionKernelCommandLine=` test that systemd offers.

bcressey commented on packages/os/kdump-tmpfiles.conf:
```diff
@@ -0,0 +1,4 @@
+d /local/boot 0700 root root -
+d /local/var/log/kdump 0700 root root -
+T /local/boot - - - - security.selinux=system_u:object_r:secret_t:s0
```
Can we just mount this at `/boot`? Putting it under `/local` seems weird.

Also the label should be `os_t` since it's not a secret.

bcressey commented on packages/os/kdump-tmpfiles.conf:
```diff
@@ -0,0 +1,4 @@
+d /local/boot 0700 root root -
+d /local/var/log/kdump 0700 root root -
```
I'd prefer to use the "normal" names for these paths:
```suggestion
d /var/log/kdump 0700 root root -
```

bcressey commented on packages/release/release-sysctl.conf:
```diff
@@ -44,9 +44,6 @@ kernel.kptr_restrict = 1
 # Avoid kernel memory address exposures via dmesg.
 kernel.dmesg_restrict = 1
 
-# Turn off kexec, even if it's built in.
-kernel.kexec_load_disabled = 1
```
What I'd like to do instead is:

1. start the system with kexec load enabled
2. (optionally) if we have a crash kernel reservation, load the kernel
3. set kexec_load_disabled = 1

Ideally this would happen in `preconfigured.target` so it runs before any container-supplied code.

bcressey commented on sources/models/shared-defaults/kexec-load-disabled.toml:
```diff
@@ -0,0 +1,3 @@
+# Turn off kexec, even if it's built in.
+[settings.kernel.sysctl]
+"kernel.kexec_load_disabled" = "1"
```
These should go away with the change in `kexec_load_disabled` handling that I suggested.

tjkirch replied:
^ I like that plan, but if it doesn't work for some reason, I think we'll want a migration to remove this sysctl setting on downgrade.

bcressey commented on packages/release/opt.mount:
```diff
@@ -5,6 +5,7 @@ Conflicts=umount.target
 Before=local-fs.target umount.target
 Wants=prepare-local.service
 After=prepare-local.service
+ConditionPathExists=!/proc/vmcore
```
nit: I'd probably leave `/opt` alone since it's just a bind mount, like `/var`.

bcressey commented on packages/release/usr-share-licenses.mount.in:
```diff
@@ -2,6 +2,7 @@
 Description=License files
 Conflicts=umount.target
 Before=local-fs.target umount.target
+ConditionPathExists=!/proc/vmcore
```
I'm OK with this, but for some of these mounts I wonder if we should just move them to happen in `preconfigured.target` where they can run in parallel with other system configuration. Then they wouldn't be automatically pulled in by `local-fs.target`, and we wouldn't need this condition.

bcressey commented on packages/os/os.spec:
```diff
@@ -227,6 +237,13 @@ Summary: Setting generator for populating admin container user-data from IMDS.
 %endif
 
+%if %{_is_vmware_variant} || %{_is_dev_variant}
+%package -n %{_cross_os}prairie-dog
+Summary: Tools for kdump support
+%description -n %{_cross_os}prairie-dog
+%{summary}.
```
`prairiedog` should depend on `kexec-tools` and `makedumpfile`, but we don't really want to delay the `os` package build waiting for those.

I would give it the same treatment we've given `host-ctr`, where we have "Requires" in the spec, and add comments to `Cargo.toml` saying how we'll pull it in via `release` instead. Then just add these tools unconditionally to `release`.

That ties into my other feedback - kdump support is really a distro feature that we'll want an easy way to turn on or off by just setting the right command line parameter in the variant definition.

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+// Used to pass None to nix::mount::mount
+const NONE: Option<&'static [u8]> = None;
+
+/// Stores user-supplied global arguments
+#[derive(Debug)]
+struct Args {
+    log_level: LevelFilter,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+        }
+    }
+}
```
You can feel free to use `argh` rather than hand-rolled argument parsing code. Size is a bit of a concern for this binary, given where it runs, but `argh` doesn't cost very much.

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+    }
+
+    info!("Generating memory dump");
+    // Extract kdump-compressed dump file, without empty pages, using
+    // zlib compression
```
I'm looking at pulling in zstd for other reasons - compressing the kernel and modules - and this is another good use case for it.

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+                subcommand = Some(arg)
+            }
+
+            // Other arguments are passed to the subcommand parser
+            _ => subcommand_args.push(arg),
```
Do we do anything with `subcommand_args` ?

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+        ],
+    )?;
+
+    info!("Generating dmesg dump");
+    // --dump-dmesg generates a dump with only dmesg logs
+    command(
+        MAKEDUMPFILE_PATH,
+        &[
+            "--dump-dmesg",
+            "--message-level",
+            "4",
+            "/proc/vmcore",
+            dmesg_file_path.as_ref(),
+        ],
+    )?;
```
I'd reverse these operations since dmesg extraction is quick, and core compression is slow. If we get rebooted before we can finish, at least we'd have some of the data.

bcressey commented on packages/kexec-tools/kexec-tools.spec:
```diff
@@ -0,0 +1,32 @@
+%license COPYING
+%{_cross_attribution_file}
+%{_cross_sbindir}/kexec
+%{_cross_sbindir}/vmcore-dmesg
```
We could exclude `vmcore-dmesg` too if we don't ever use it.

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+    }
+
+    info!("Loading crash kernel");
+    // Load the panic kernel from `BOOT_MOUNT_PATH`, letting kexec decide which syscall
+    // it should it use (KEXEC_LOAD and KEXEC_FILE_LOAD).
```
Can we force it to always use `KEXEC_FILE_LOAD`?

arnaldo2792 replied:
Yes we can, but for some reason the 5.4 kernel fails to use the `kexec_file_load` syscall, and it returns an `ENOSUP` error. The problem is not in the `kexec` library but rather in the kernel, so I need to debug the syscall to check what's returning the error. I'm already working with the kernel folks on this. 

In the 5.10 kernel, kexec always uses `kexec_file_load`, so I think it is OK to use `-a` for now and let `kexec` decide what syscall to use, since we are only enabling kdump in variants that use the 5.10 kernel.

tjkirch replied:
Wouldn't we expect to enable this in other variants relatively soon?  I imagine people building their own images/variants might want to enable it as well.  Wouldn't want a nasty surprise.

arnaldo2792 replied:
Even if we enable this for other variants, and people build their own images/variants, they will still be locked with the `kexec_load` syscall for the aarch64 5.4 kernel variants since the syscall for this aarch64 was implemented after the 5.4 version.

bcressey commented on sources/prairie-dog/src/main.rs:
```diff
@@ -0,0 +1,431 @@
+            "-pa",
+            "--reuse-cmdline",
+            "--append",
+            "irqpoll maxcpus=1 reset_devices systemd.unit='kdump.service' udev.children-max=1",
```
`irqpoll` isn't one of the recommended flags for arm64

arnaldo2792 replied:
I updated the cmdline parameters, and now we are using what AL2 uses, plus the recommended parameters.

arnaldo2792 committed:
systemd: move systemd mounts to preconfigured.target

This commit moves some of the systemd mount units to be part of
`preconfigured` instead of the `local-fs`. This is to reduce the
overhead of mounting unnecesary mount endpoints during the execution of
the crash kernel.

arnaldo2792 commented:
Changes in force push:
- Update `kexec-tools`, `makedumpfile`, and `libelf` to latest release
- Rename `prairie-dog` to `prairiedog`
- Remove conditionals in `os.spec`, and build all the added packages regardless of the variant
- Add the `disable-kexec-load.service` systemd unit to always disable `kexec*` syscalls regardless of the state of `kexec.service`
- Only execute `kexec.service` if there is memory reserved for the crash kernel
- Update cmdline parameters for crash kernel, and use the same parameters as Amazon Linux 2
- Add `irqpoll` to the cmdline for the crash kernel, depending on the target architecture
- Move unnecessary mounts from `local-fs.target` to `preconfigured.target`

arnaldo2792 commented:
Force push updates:

-  Gracefully exit while loading the crash kernel when no memory was reserved for the crash kernel

arnaldo2792 committed:
kernel-5.4: add patch required for kdump support

bcressey commented on packages/libelf/libelf.spec:
```diff
@@ -0,0 +1,56 @@
+%files devel
+%{_cross_includedir}/*.h
+%{_cross_includedir}/elfutils/*.h
+%{_cross_libdir}/*.so
```
These need to be more narrowly scoped so we don't include files that are in the main package.
```suggestion
%{_cross_libdir}/libasm.so
%{_cross_libdir}/libelf.so
%{_cross_libdir}/libdw.so
```

bcressey commented on packages/makedumpfile/0000-fix-makefile-to-allow-cross-compilation.patch:
```diff
@@ -0,0 +1,25 @@
+
+---
+ Makefile | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
```
I'm surprised we need this; the upstream code dates from the commit that added cross-compilation support.

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+const BOOT_MOUNT_PATH: &str = "/boot";
+
+// Mounts configurations
+const BOOT_MOUNT_SELINUX_CONTEXT: &str = "context=system_u:object_r:os_t:s0";
```
We should label the boot filesystem in `rpm2img` rather than relying on a context option.

bcressey commented on packages/os/kdump-tmpfiles.conf:
```diff
@@ -0,0 +1,2 @@
+d /var/log/kdump 0700 root root -
+T /var/log/kdump - - - - security.selinux=system_u:object_r:os_t:s0
```
```suggestion
d /var/log/kdump 0700 root root -
T /var/log/kdump - - - - security.selinux=system_u:object_r:secret_t:s0
```

bcressey commented on packages/os/configure-boot-mount.service:
```diff
@@ -0,0 +1,16 @@
+RefuseManualStart=true
+RefuseManualStop=true
+Wants=prepare-local.service
+After=prepare-local.service
```
We don't depend on `/local` any more.

I'd like to move this to the `release` package and align the unit name and description with `prepare-local.service`, i.e. `prepare-boot.service`. Maybe add a comment to indicate that finding the boot partition is not trivial and that's why we need the helper program.

```suggestion
[Unit]
Description=Prepare Boot Directory (/boot)
```

bcressey commented on packages/os/configure-boot-mount.service:
```diff
@@ -0,0 +1,16 @@
+StandardError=journal+console
+
+[Install]
+WantedBy=local-fs.target
```
Seems more consistent to give this the same treatment as other mount units:
```suggestion
[Install]
WantedBy=preconfigured.target
```

bcressey commented on packages/os/disable-kexec-load.service:
```diff
@@ -0,0 +1,17 @@
+Type=oneshot
+# We could set this setting using the API, however, doing so
+# requires us to add additional dependencies to this simple
+# oneshot service
```
I find this comment a bit misleading. It isn't just a matter of dependencies - the direct route is better here since it's a security property we're trying to enforce.

bcressey commented on packages/os/disable-kexec-load.service:
```diff
@@ -0,0 +1,17 @@
+StandardError=journal+console
+
+[Install]
+WantedBy=preconfigured.target
```
```suggestion
[Install]
RequiredBy=preconfigured.target
```

bcressey commented on packages/os/kexec.service:
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Load crash kernel
+ConditionKernelCommandLine=crashkernel
+ConditionPathExists=!/proc/vmcore
```
`kexec.service` is a bit generic - maybe `load-crash-kernel.service` to align with subcommand?

bcressey commented on packages/os/configure-boot-mount.service:
```diff
@@ -0,0 +1,16 @@
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/prairiedog configure-boot-mount
```
Maybe this for consistency with the suggested unit name:
```suggestion
ExecStart=/usr/bin/prairiedog prepare-boot
```

bcressey commented on packages/os/kdump.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Create dump after kernel crash
```
`kdump.service` is a bit generic - at a quick glance I'd expect this to be the service that loads the crash kernel.

Maybe `capture-kernel-dump.service`, and align the name with the prairiedog subcommand.

```suggestion
[Unit]
Description=Capture dump after kernel crash
```

bcressey commented on packages/os/os.spec:
```diff
@@ -37,13 +37,18 @@ Source111: metricdog.service
+Source115: kdump.service
+Source116: configure-boot-mount.service
+Source117: kexec.service
+Source118: disable-kexec-load.service
```
Let's put all these files into "release", which is roughly for files and services that you'd expect on a normal Linux distro, versus API related functionality.

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+const KDUMP_LOGS_PATH: &str = "/var/log/kdump";
+const LOG_FILE: &str = "prairiedog.log";
+const DMESG_DUMP_FILE: &str = "dmesg.dump";
+const KDUMP_FILE: &str = "kdump.dump";
```
To eliminate some of the redundancy:
```suggestion
const KDUMP_FILE: &str = "vmcore.dump";
```

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+// Files generated by prairiedog
+const KDUMP_LOGS_PATH: &str = "/var/log/kdump";
+const LOG_FILE: &str = "prairiedog.log";
+const DMESG_DUMP_FILE: &str = "dmesg.dump";
```
Since it's plain text:
```suggestion
const DMESG_DUMP_FILE: &str = "dmesg.log";
```

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+// Mounts the active boot partition
+fn configure_boot_mount() -> Result<()> {
+    // Get the current partitions state
+    let state = State::load().context(error::LoadState)?;
```
"State" is pretty generic:
```suggestion
    let state = signpost::State::load().context(error::LoadState)?;
```

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+        path: BOOT_MOUNT_PATH,
+    })?;
+
+    // Make the mount point private so new mount namespaces don't have
+    // access to it. This has to be set as a different call otherwise
+    // the mount syscall returns `EINVAL`
+    nix::mount::mount(
+        NONE,
+        BOOT_MOUNT_PATH,
+        NONE,
+        nix::mount::MsFlags::MS_PRIVATE,
+        NONE,
+    )
```
We should be able to specify "private" as a mount option, right?

arnaldo2792 replied:
I tried to add `private` as an option, but the syscall failed

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+    // Load the panic kernel from `BOOT_MOUNT_PATH`, letting kexec decide which syscall
+    // it should it use (KEXEC_LOAD and KEXEC_FILE_LOAD). We will use this setting until
+    // we figure out why kexec doesn't recognize the 5.4 kernel image as a valid file
+    // to be used with KEXEC_FILE_LOAD.
```
If we're planning to merge this prior to a fix, let's open an issue to track it.

arnaldo2792 replied:
I got to port the kernel patch to make the `kexec_file_load` syscall work in the x86_64 5.4 kernel. However, I'm still waiting for the kernel folks to help me out with porting the syscall for aarch64, since the syscall was introduced for aarch64 on a higher kernel version.
 
I talked to @tjkirch since he expressed his concerns of using different syscalls depending on the kernel version/architecture. We agreed that we can proceed with my PR provided that we document in an GH issue that we will force the `kexec_file_load` syscall once the kernel folks add support for it for the 5.4 kernel.

tjkirch replied:
I think we should document the restrictions in the same place we document the feature, not (only) in a GH issue.

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+            command: Command,
+            source: std::io::Error,
+        },
+        #[snafu(display("Kexec load syscalls are disabled, please make sure the value of `kernel.kexec_load_disabled` is 0"))]
```
```suggestion
        },
        
        #[snafu(display("Kexec load syscalls are disabled, please make sure the value of `kernel.kexec_load_disabled` is 0"))]
```

bcressey commented on sources/prairiedog/src/main.rs:
```diff
@@ -0,0 +1,359 @@
+        #[snafu(display("Failed to load partitions state: {}", source))]
+        LoadState { source: SignpostError },
+
+        #[snafu(display("Logger setup error: {}", source))]
```
```suggestion
        #[snafu(display("Failed to set up logger: {}", source))]
```

arnaldo2792 commented:
Forced push includes:

- 5.4 kernel patch to use `kexec_file_load` with the 5.4 kernel
- makedumpfile patch renamed
- Renamed `kexec.service`, `kdump.service` and `configure-boot-mount.service` to `load-crash-kernel.service`, `capture-kernel-dump.service`, and `prepare-boot.service`
- Renamed prairiedog command to `prepare-boot.service`
- Moved all systemd services created in this PR to the release package
- Label boot partition in `rpm2img`
- Narrow down the libraries include in the `libelf-devel`package

bcressey commented on packages/makedumpfile/0000-fix-strip-invocation-for-TARGET-env-variable.patch:
```diff
@@ -0,0 +1,25 @@
+ # Use TARGET as the target architecture if specified.
+ # Defaults to uname -m
+-ifeq ($(strip($TARGET)),)
++ifeq ($(strip $TARGET),)
```
I am pretty sure the stripped value ends up as "ARGET" meaning this would always be false.

```suggestion
+ifeq ($(strip $(TARGET)),)
```

bcressey commented on packages/os/kdump-tmpfiles.conf:
```diff
@@ -0,0 +1,2 @@
+d /var/log/kdump 0700 root root -
+T /var/log/kdump - - - - security.selinux=system_u:object_r:secret_t:s0
```
nit: could put this in `release-tmpfiles.conf` instead

bcressey commented on packages/release/release.spec:
```diff
@@ -38,6 +38,14 @@ Source1020: var-lib-kernel-devel-lower.mount.in
 Source1021: usr-src-kernels.mount.in
 Source1022: usr-share-licenses.mount.in
 
+# Mounts that require helper programs
+Source1040: prepare-boot.service
```
nit: should we move `prepare-local.service` down here too?

arnaldo2792 commented:
Forced push includes:

- Remove unnecessary patch in the 5.4 kernel that caused failures in the aarch64 variants' build 
- Update commit messages

arnaldo2792 commented:
Forced push includes:

- Added documentation to README
- Fixed makedumpfile patch
- Addressed suggested nit changes

bcressey commented on README.md:
```diff
@@ -583,6 +583,27 @@ ssh -i YOUR_KEY_FILE \
+
+Bottlerocket provides support to collect kernel crash dumps whenever the system kernel panics.
+Like other Linux distributions, Bottlerocket uses the system kernel as the crash kernel to keep the image size small.
+Unlike other distros, Bottlerocket doesn't require a separate initfamfs file to load the crash kernel.
```
I'd prefer to omit these; there's not a lot of value in comparing our implementation to other distros.

bcressey commented on README.md:
```diff
@@ -583,6 +583,27 @@ ssh -i YOUR_KEY_FILE \
+Like other Linux distributions, Bottlerocket uses the system kernel as the crash kernel to keep the image size small.
+Unlike other distros, Bottlerocket doesn't require a separate initfamfs file to load the crash kernel.
+Once the system panics, both the dmesg log and vmcore dump are stored at `/var/log/kdump`.
+The following pages are excluded from the collected vmcore dump:
+
+* Pages filled with zero
+* Non-private cache pages
+* All cache pages
+* User process data pages
+* Free pages
```
This also seems like too much information.

bcressey commented on README.md:
```diff
@@ -583,6 +583,27 @@ ssh -i YOUR_KEY_FILE \
+
+* The system kernel will reserve 256M for the crash kernel, only when the host has at least 2GB of memory; the reserved space won't be available for processes running in the host
+* The crash kernel will only be loaded when the `crashkernel` parameter is present in the kernel's cmdline and if there is memory reserved for it
+* For the time being, only the vmware and aws-dev variants have kdump support enabled
+* For the time being, variants built for aarch64 and that use the 5.4 kernel will use the `kexec_load` syscall instead of the newer `kexec_file_load` syscall
```
I'd omit both of these notes; we don't talk much about the `*-dev` variants elsewhere.

We could say "Currently, only vmware variants have kdump support enabled."

arnaldo2792 commented:
Forced push include fixes in README

jpculp commented on README.md:
```diff
@@ -583,6 +583,17 @@ ssh -i YOUR_KEY_FILE \
+There area few important caveats about the provided kdump support:
+
+* Currently, only vmware variants have kdump support enabled
+* The system kernel will reserve 256M for the crash kernel, only when the host has at least 2GB of memory; the reserved space won't be available for processes running in the host
```
Nit. I would change 256M to 256MB to be consistent with "at least 2GB of memory".

arnaldo2792 commented:
Forced push includes:
- Patches to make the `kexec_file_load` syscall work in arm64
- Fixed nit in README
- Force the `kexec_file_load` syscall in prairedog

bcressey commented on README.md:
```diff
@@ -583,6 +583,17 @@ ssh -i YOUR_KEY_FILE \
+Bottlerocket provides support to collect kernel crash dumps whenever the system kernel panics.
+Once this happens, both the dmesg log and vmcore dump are stored at `/var/log/kdump`, and the system reboots.
+
+There area few important caveats about the provided kdump support:
```
```suggestion
There are a few important caveats about the provided kdump support:
```

arnaldo2792 commented:
Forced pushed includes nit fix
